File: python/collatz-conjecture/collatz_conjecture.py
def steps(number):
    if number <= 0:
        raise ValueError("invalid number")
    else:
        steps = 0
        while number != 1:
            steps += 1
            # odd number
            if bool(number & 1):
                number = (number * 3) + 1
            # even number
            else:
                number = int(number / 2)
    return steps


def version():
    VERSION = 1

File: ruby/collatz-conjecture/collatz_conjecture.rb
# frozen_string_literal: true

require 'pry'

class CollatzConjecture
  class << self
    def steps(number)
      raise ArgumentError unless valid_number?(number)
      return 0 if number == 1

      count_steps(number)
    end

    private

    def valid_number?(number)
      number.positive?
    end

    def count_steps(number)
      steps = 0
      while number != 1
        steps += 1
        if number.even?
          number /= 2
        else
          number = (3 * number) + 1
        end
      end
      steps
    end
  end
end

module Version
  VERSION = 1
end
